### ISSUE TITLE
Fitting resamples on a stack

### ISSUE THREAD

**Issue by Jeffrothschild** [tidymodels/stacks#130]
< -- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ -->

## Fit resamples on a model stack object

Could there be a way do the equivalent of fit_resamples() when using a stacks? fit_resamples doesn't work with a stack, but I would love to be able to collect metrics on bootstrapped resamples.

For example, after getting the tuned result ....

```
individ_stack <- 
  stacks() %>% 
  add_candidates(tune_rs)

individ_ens <- blend_predictions(individ_stack)

resamp_res <- fit_resamples(
  individ_ens,
  validation_boots,
  control = control_resamples(save_pred = TRUE)
)

resampled_preds <- collect_predictions(resamp_res, summarize = T)

```
Thanks!

**Comment by simonpcouch** [comment]
Thanks for the issue!

From my understanding of your issue, stacks already does the thing you propose—under the hood, `blend_predictions` is fitting the meta-learner to bootstrap resamples of the data stack:

https://github.com/tidymodels/stacks/blob/9f529cb/R/blend_predictions.R#L203-L215

If this is not what you mean, can you please provide a minimal [reprex](https://github.com/tidyverse/reprex#what-is-a-reprex) (reproducible example) demonstrating the (lack of) functionality you're proposing?

**Comment by Jeffrothschild** [comment]
Hi Simon,

Thanks for your reply. I think I'm still looking for something different, more related to what happens after the stack has been fit. 

We would normally predict on new (test) data and evaluate the metrics based on that. There are times when I don't have enough data to make a test/train split, but would like to use stacking. In this case the ensemble metrics are obviously going to be overly optimistic. It seems like the best option then would be to get metrics on bootstrap resamples.

In the example below, I'm wondering if there is a way to get the `ens_metrics` output below to be arrived at in a way similar to the `resampled_metrics` output at the bottom (achieved using a single model approach). 

Hopefully that makes sense, I'm happy to clarify further.

Thanks, I can't tell you how much I appreciate this package!


```

library("DALEX")
library("dplyr")
library("tidymodels")
library("stacks")
tidymodels_prefer()


fifa_small <- fifa %>% 
  select(value_eur, age, 
         attacking_crossing:attacking_volleys, 
         defending_marking:defending_sliding_tackle) %>% 
  as_tibble() %>% dplyr::slice_sample(n = 1000)


fifa_small_folds <- vfold_cv(fifa_small, v = 8, repeats = 1)

basic_rec <-   
  recipe(value_eur ~ ., data = fifa_small) %>%
  step_nzv(all_numeric_predictors()) %>% 
  step_normalize(all_numeric(), -all_outcomes()) 

model1 <- 
  boost_tree(trees = 1000) %>%
  set_engine('xgboost', importance = TRUE) %>%
  set_mode('regression')

model2 <-
  linear_reg(penalty = tune(), mixture = 1) %>%
  set_engine('glmnet')

model3 <-
  linear_reg(penalty = tune(), mixture = 0) %>%
  set_engine('glmnet')

wfs <- 
  workflow_set(
    preproc = list(basic_rec),  
    models = list(model1, model2, model3), 
    cross = T  )

wfs


wfs_rs <-
  workflow_map(
    wfs,
    "tune_grid",
    resamples = fifa_small_folds,
    grid = 10,
    control = control_grid(save_pred = TRUE,
                           parallel_over = "everything",
                           save_workflow = TRUE    ) )

wfs_rs

wfs_stack <- 
  stacks() %>% 
  add_candidates(wfs_rs)


blend_ens <- blend_predictions(wfs_stack, penalty = 10^seq(-2, 0, length = 10))
blend_ens

ens_fit <- fit_members(blend_ens)

# get metrics based on the ens fit on the training set 
ens_preds <- predict(ens_fit, fifa_small) %>% 
  bind_cols(fifa_small) 

ens_preds %>% metrics(value_eur, .pred)

ens_metrics <- ens_preds %>% metrics(value_eur, .pred)
ens_metrics

##### non-stacking option - 

best_id <- wfs_rs %>%  
  rank_results(
    rank_metric = "rmse",
    select_best = TRUE
  ) %>% 
  dplyr::slice(1) %>% 
  pull(wflow_id)


## Get the best workflowset
wf_best <- extract_workflow(wfs_rs, id = best_id)

## extract the tuned results from the best workflow
wf_best_tuned <- wfs_rs[wfs_rs$wflow_id == best_id,
                                          "result"][[1]][[1]]

#see metrics on best tuning result
best_tuned_metrics <- collect_metrics(wf_best_tuned)
best_tuned_metrics


best_final <- finalize_workflow(wf_best, select_best(wf_best_tuned, "rmse"))

final_fit <- best_final %>% fit(fifa_small)

# see metrics based on the final fit on the training set
preds <- predict(final_fit, fifa_small) %>%  bind_cols(fifa_small) 
preds %>% metrics(value_eur, .pred)


#use fit resamples to see performance
boots <-  bootstraps(fifa_small, times = 50)

resampled <- fit_resamples(
  best_final,
  boots,
  control = control_resamples(save_pred = TRUE)
)

resampled_metrics <- collect_predictions(resampled, summarize = T) %>% 
  metrics(value_eur, .pred)

resampled_metrics

**Comment by simonpcouch** [comment]
Thanks for taking the time to put this together! I think I have a better sense of what you're getting at now.

> We would normally predict on new (test) data and evaluate the metrics based on that. There are times when I don't have enough data to make a test/train split, but would like to use stacking. In this case the ensemble metrics are obviously going to be overly optimistic. 

Yup, this is true.

> It seems like the best option then would be to get metrics on bootstrap resamples.

A trained model stack is similar to other parsnip model objects in that, given the same set of predictors, the stack will make the same prediction. Thus, taking bootstrap resamples of the training data—or any resampling strategy where whole rows of the input data are preserved—_after_ the model fit and assessing model performance by having that model fit predict against the resamples would not help with overfitting/re-predicting on data the model has "seen." Predicting on bootstraps will still make the model fit seem overly performant.

I think stacks will continue not accommodating this kind of workflow. A train/test split is simply necessary to assess model performance, in this case. In cases where `n` is a constraint, one ought to consider whether model stacking is an appropriate approach, and, if so, could allot a larger portion of the data to training than the default `prop = 3/4`. 

> In the example below, I'm wondering if there is a way to get the ens_metrics output below to be arrived at in a way similar to the resampled_metrics output at the bottom 

The `resampled_metrics` approach seems to be proposing something a bit different: fitting a model specification to resamples of the training set, collecting predictions from the resampled fits, and then evaluating performance of the specification fitted to the entire training set by comparing predictions summarized across the resamples against a validation set of those resamples. 

{stacks} does this already in the sense where "the training set" is the `data_stack`, at the `blend_predictions` step. An analogue to `fit_resamples` in place of `fit_resamples` does not address the problem that you later solve with the single-model approach—`blend_predictions` has estimated the meta-learner's coefficients already, and `fit_members` does not alter _those_ coefficients. (i.e. `fit_members` does not change the weights with which member's predictions are combined—it changes the predictions from the members themselves. The coefficients of the model fitted in `blend_predictions` map to columns in the `data_stack`; those in `fit_members` map to columns in the original training data.) Introducing another layer of resampling at `blend_predictions` or before would be redundant with the current resampling schemes.

**Comment by Jeffrothschild** [comment]
Thanks Simon, I appreciate the clarification!
